contracts/contracts/vault/OETHVaultCore.sol

```diff
@@ -125,24 +137,338 @@ contract OETHVaultCore is VaultCore {
             "Redeem amount lower than minimum"
         );
 
-        if (IERC20(weth).balanceOf(address(this)) >= amountMinusFee) {
-            // Use Vault funds first if sufficient
-            IERC20(weth).safeTransfer(msg.sender, amountMinusFee);
-        } else {
-            address strategyAddr = assetDefaultStrategies[weth];
-            if (strategyAddr != address(0)) {
-                // Nothing in Vault, but something in Strategy, send from there
-                IStrategy strategy = IStrategy(strategyAddr);
-                strategy.withdraw(msg.sender, weth, amountMinusFee);
-            } else {
-                // Cant find funds anywhere
-                revert("Liquidity error");
-            }
-        }
+        // Is there enough WETH in the Vault available after accounting for the withdrawal queue
+        require(_wethAvailable() >= amountMinusFee, "Liquidity error");
+
+        // Transfer WETH minus the fee to the redeemer
+        IERC20(weth).safeTransfer(msg.sender, amountMinusFee);
 
         // Burn OETH from user (including fees)
         oUSD.burn(msg.sender, _amount);
 
+        // Prevent insolvency
         _postRedeem(_amount);
     }
+
+    /**
+     * @notice Request an asynchronous withdrawal of WETH in exchange for OETH.
+     * The OETH is burned on request and the WETH is transferred to the withdrawer on claim.
+     * This request can be claimed once the withdrawal queue's `claimable` amount
+     * is greater than or equal this request's `queued` amount.
+     * There is no minimum time or block number before a request can be claimed. It just needs
+     * enough WETH liquidity in the Vault to satisfy all the outstanding requests to that point in the queue.
+     * OETH is converted to WETH at 1:1.
+     * @param _amount Amount of OETH to burn.
+     * @param requestId Unique ID for the withdrawal request
+     * @param queued Cumulative total of all WETH queued including already claimed requests.
+     */
+    function requestWithdrawal(uint256 _amount)
+        external
+        whenNotCapitalPaused
+        nonReentrant
+        returns (uint256 requestId, uint256 queued)
+    {
+        // Burn the user's OETH
+        // This also checks the requester has enough OETH to burn
+        oUSD.burn(msg.sender, _amount);
```

Comment:
Let's move this burn down to just before the post redeem. Won't actually change anything, but it's better practice since:

1. Let's us do internal calcs, storage updates first, then external calls, then checks, which is a great ordering.
2. It's consistent with the redeem method
2. Even though there are no reentrancy issues here, I like to do updates in a safe order. For OUSD/OETH, by lowering the vault value first, before reducing total supply, there's never a time when the state makes the vault look like it has money to hand out. If doing this made the code look ugly, we could skip, but it makes the code look better I think. We can leave a comment at the start of the method about the user's balance being checked by the later burn.

Reply:
I've moved the burn down to before the post redeem.

I also moved the emit of the `WithdrawalRequested` event to the bottom. Most of the time we emit events at the end but mint and redeem does them first for some reason.